validator/server_arb/execution_run_test.go

```diff
@@ -0,0 +1,208 @@
+	return nil
+}
+func (m *mockMachine) Freeze()  {}
+func (m *mockMachine) Destroy() {}
+
+func Test_machineHashesWithStep(t *testing.T) {
+	mm := &mockMachine{}
+	e := &executionRun{}
+	ctx := context.Background()
+
+	t.Run("basic argument checks", func(t *testing.T) {
+		machStartIndex := uint64(0)
+		stepSize := uint64(0)
+		numRequiredHashes := uint64(0)
+		_, err := e.machineHashesWithStepSize(ctx, machStartIndex, stepSize, numRequiredHashes)
+		if !strings.Contains(err.Error(), "step size cannot be 0") {
+			t.Fatal("Wrong error")
+		}
+		stepSize = uint64(1)
+		_, err = e.machineHashesWithStepSize(ctx, machStartIndex, stepSize, numRequiredHashes)
+		if !strings.Contains(err.Error(), "required number of hashes cannot be 0") {
+			t.Fatal("Wrong error")
+		}
+	})
+	t.Run("machine at start index 0 hash is the finished state hash", func(t *testing.T) {
+		mm.gs = validator.GoGlobalState{
+			Batch: 1,
+		}
+		machStartIndex := uint64(0)
+		stepSize := uint64(1)
+		numRequiredHashes := uint64(1)
+		e.cache = &MachineCache{
+			buildingLock: make(chan struct{}, 1),
+			machines:     []MachineInterface{mm},
+			finalMachine: mm,
+		}
+		go func() {
+			<-time.After(time.Millisecond * 50)
+			e.cache.buildingLock <- struct{}{}
+		}()
+		hashes, err := e.machineHashesWithStepSize(ctx, machStartIndex, stepSize, numRequiredHashes)
+		if err != nil {
+			t.Fatal(err)
+		}
+		expected := machineFinishedHash(mm.gs)
+		if len(hashes) != 1 {
+			t.Fatal("Wanted one hash")
+		}
+		if expected != hashes[0] {
+			t.Fatalf("Wanted %#x, got %#x", expected, hashes[0])
+		}
+	})
+	t.Run("can step in step size increments and collect hashes", func(t *testing.T) {
+		initialGs := validator.GoGlobalState{
+			Batch:      1,
+			PosInBatch: 0,
+		}
+		mm.gs = initialGs
+		mm.totalSteps = 20
+		machStartIndex := uint64(0)
+		stepSize := uint64(5)
+		numRequiredHashes := uint64(4)
+		e.cache = &MachineCache{
+			buildingLock: make(chan struct{}, 1),
+			machines:     []MachineInterface{mm},
+			finalMachine: mm,
+		}
+		go func() {
+			<-time.After(time.Millisecond * 50)
+			e.cache.buildingLock <- struct{}{}
+		}()
+		hashes, err := e.machineHashesWithStepSize(ctx, machStartIndex, stepSize, numRequiredHashes)
+		if err != nil {
+			t.Fatal(err)
+		}
+		expectedHashes := make([]common.Hash, 0)
+		for i := uint64(0); i < 4; i++ {
+			if i == 0 {
+				expectedHashes = append(expectedHashes, machineFinishedHash(initialGs))
+				continue
+			}
+			gs := validator.GoGlobalState{
+				Batch:      1,
+				PosInBatch: uint64(i * stepSize),
+			}
+			expectedHashes = append(expectedHashes, gs.Hash())
+		}
+		if len(hashes) != len(expectedHashes) {
+			t.Fatal("Wanted one hash")
+		}
+		for i := range hashes {
+			if expectedHashes[i] != hashes[i] {
+				t.Fatalf("Wanted at index %d, %#x, got %#x", i, expectedHashes[i], hashes[i])
+			}
+		}
+	})
+	t.Run("if finishes execution early, simply pads the remaining desired hashes with the machine finished hash", func(t *testing.T) {
```

Comment:
Well, this answers my earlier question. Why is this the protocol? Why not just return fewer hashes if the machine finishes? Seems like it *could* reduce the size of the data passed back to the callers?

Reply:
Agreed...will let the caller pad if they need to